Add command to generate sitemaps using command line

deefco opened the pull request:
<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

This pull request adds a console command sitemap:generate that generates all configured sitemaps in the backend. To achieve this "cleanly" several things have been added, first Sitemap model has been interfaced using SitemapInterface, fairly clean interface, only generateXml() is not a dto operation. Secondly SitemapRepository and SitemapRepositoryInterface was added to seperate crud operations from model.

Console command simply does sitemaprepository->getList() and iterates over the result, calling generateXml() on each one. It seems to me that generateXml should be moved to an spi contract, SitemapXmlGenerator::execute() or something. However it also seems to me that if we go down that route this pull request will never get merged, so I'd like to do that at a later date.

Bonus question: is there a way to get console components like progressbar without either object manager or calling new?

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. https://github.com/magento/community-features/issues/67 


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

deefco committed:
Add sitemap generate command, add sitemap interface, add sitemap repository, add interfaces for all this stuff

deefco committed:
Add sitemap search results + getlist command in repository

deefco committed:
Make command generate all sitemaps and use progressbar for pretty output

magento-engcom-team commented:
Hi @deefco. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

ishakhsuvarov commented on app/code/Magento/Sitemap/Api/Data/SitemapInterface.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return $this
+     */
+    public function generateXml();
```
Data interfaces should not have any methods performing actions. It's best to move this method to a separate service

deefco replied:
Removed

ishakhsuvarov commented on app/code/Magento/Sitemap/Api/Data/SitemapInterface.php:
```diff
@@ -0,0 +1,116 @@
+/**
+ * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
+ */
+interface SitemapInterface
```
It would make sense to add support for extension attributes to this interface.

deefco replied:
Created SitemapExtensionInterface + all related things and added them

ishakhsuvarov commented on app/code/Magento/Sitemap/Api/Data/SitemapInterface.php:
```diff
@@ -0,0 +1,116 @@
+ * See COPYING.txt for license details.
+ */
+
+declare(strict_types=1);
```
Please add return types to all getters

deefco replied:
Added nullable return types to all getters

ishakhsuvarov commented on app/code/Magento/Sitemap/Api/SitemapRepositoryInterface.php:
```diff
@@ -0,0 +1,60 @@
+     * @param int $sitemapId
+     * @return SitemapInterface
+     */
+    public function getById($sitemapId): SitemapInterface;
```
`get` may be sufficient in this case

ishakhsuvarov replied:
Please add type hints to the method

ishakhsuvarov commented on app/code/Magento/Sitemap/Api/SitemapRepositoryInterface.php:
```diff
@@ -0,0 +1,60 @@
+     *
+     * @param int $sitemapId
+     * @return SitemapInterface
+     */
```
Please define exception type thrown

deefco replied:
Added

deefco committed:
Changes

deefco committed:
Use legacy xml generator, fix return types, make sitemap extensible

deefco committed:
Ran php-cs-fixer

deefco commented:
I've added extension attribute support to sitemap now, also I removed generateXml from the interface and just call generateXml on the model. Followup pull request would be splitting xml generator to spi layer but it the model needs a lot of refactoring which would be easier when it is properly interfaced. So I'm considering spi changes to be out of scope for this pr.

deefco committed:
Made sitemap extend AbstractExtensibleModel

magento-engcom-team commented:
Hi @deefco. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @deefco, here is your [new Magento instance](https://pr-18058.engcom.dev.magento.com).
Admin access: [https://pr-18058.engcom.dev.magento.com/admin](https://pr-18058.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`

deefco commented:
@magento-engcom-team give me 2.3 instance

deefco committed:
Remove explicit setid getid calls

deefco commented:
@magento-engcom-team give me 2.3-develop instance

magento-engcom-team commented:
Hi @deefco. Thank you for your request. I'm working on Magento 2.3-develop instance for you

magento-engcom-team commented:
Hi @deefco, here is your [Magento instance](https://i-18058-2-3-develop.engcom.dev.magento.com).
Admin access: [https://i-18058-2-3-develop.engcom.dev.magento.com/admin](https://i-18058-2-3-develop.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

deefco committed:
Make getStoreId nullable

deefco commented:
Could someone please take a look at this? @sidolov @ishakhsuvarov

sidolov commented on app/code/Magento/Sitemap/Api/Data/SitemapExtensionInterface.php:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Sitemap Extension interface
+ */
+interface SitemapExtensionInterface extends \Magento\Framework\Api\ExtensionAttributesInterface
```
Please, remove this interface, it will be generated automatically.

deefco replied:
Removed

sidolov commented on app/code/Magento/Sitemap/Api/Data/SitemapInterface.php:
```diff
@@ -0,0 +1,121 @@
+
+namespace Magento\Sitemap\Api\Data;
+
+interface SitemapInterface extends SitemapExtensionInterface
```
Please, extend current interface from \Magento\Framework\Api\ExtensionAttributesInterface

deefco replied:
Made it extend the right thing

sidolov commented on app/code/Magento/Sitemap/Api/Data/SitemapInterface.php:
```diff
@@ -0,0 +1,121 @@
+     * @param string $sitemapFileName
+     * @return string
+     */
+    public function getSitemapUrl($sitemapPath, $sitemapFileName): ?string;
```
Remove `Sitemap` word from the all setters/getter

sidolov replied:
Setter is missing for url

deefco replied:
getSitemapUrl was not plain dto logic so it has been removed from the interface.

sidolov commented on app/code/Magento/Sitemap/Api/SitemapRepositoryInterface.php:
```diff
@@ -0,0 +1,60 @@
+     * @throws NoSuchEntityException
+     * @return SitemapInterface
+     */
+    public function get($sitemapId): SitemapInterface;
```
Please, specify the type for `$sitemapId`

sidolov commented on app/code/Magento/Sitemap/Console/Command/GenerateCommand.php:
```diff
@@ -0,0 +1,80 @@
+
+class GenerateCommand extends Command
+{
+
```
Please, fix code style issues for the whole file

sidolov commented on app/code/Magento/Sitemap/Console/Command/GenerateCommand.php:
```diff
@@ -0,0 +1,80 @@
+        foreach ($sitemaps->getItems() as $sitemap) {
+            $progressbar->setMessage('Generating sitemap ' . $sitemap->getId() . ' ...');
+            $progressbar->display();
+            $sitemap->generateXml();
```
It would be great to introduce error handling for the generation flow

sidolov commented on app/code/Magento/Sitemap/Model/ResourceModel/SitemapRepository.php:
```diff
@@ -0,0 +1,125 @@
+
+class SitemapRepository implements SitemapRepositoryInterface
+{
+
```
Please, fix code style issues for the whole file

sidolov commented on app/code/Magento/Sitemap/Model/ResourceModel/SitemapRepository.php:
```diff
@@ -0,0 +1,125 @@
+        $searchResults->setSearchCriteria($searchCriteria);
+
+        /** @var \Magento\Sitemap\Model\ResourceModel\Sitemap\Collection $collection */
+        $collection = $this->sitemapFactory->create()->getCollection();
```
Please, use collection factory in this case

sidolov commented on app/code/Magento/Sitemap/Model/ResourceModel/SitemapRepository.php:
```diff
@@ -0,0 +1,125 @@
+        $sitemaps = [];
+        /** @var SitemapInterface $sitemap */
+        foreach ($collection->getItems() as $sitemap) {
+            $sitemaps[] = $this->get($sitemap->getId());
```
Please clarify, if this additional requests really needed

deefco replied:
This is the way it is done in BookmarksRepository in the core, please point me to correct implementation if this is wrong way to do it?

sidolov commented on app/code/Magento/Sitemap/Model/ResourceModel/SitemapRepository.php:
```diff
@@ -0,0 +1,125 @@
+    public function delete(SitemapInterface $sitemap): bool
+    {
+        try {
+            $this->sitemapResource->save($sitemap);
```
Looks like `save` method called by mistake. Please, use `delete`

deefco replied:
Fixed

sidolov commented on app/code/Magento/Sitemap/Model/Sitemap.php:
```diff
@@ -192,30 +187,36 @@ class Sitemap extends \Magento\Framework\Model\AbstractModel implements \Magento
-     * @param DocumentRoot|null $documentRoot
-     * @param ItemProviderInterface|null $itemProvider
-     * @param SitemapConfigReaderInterface|null $configReader
+     * @param \Magento\Framework\Model\Context                        $context
```
Please, revert changes for doc block, Magento static tests will be failed due to specific coding standards

deefco replied:
Doc block has been changed because class inheritance has been changed.

sidolov commented on app/code/Magento/Sitemap/Model/Sitemap.php:
```diff
@@ -192,30 +187,36 @@ class Sitemap extends \Magento\Framework\Model\AbstractModel implements \Magento
         \Magento\Framework\Model\Context $context,
         \Magento\Framework\Registry $registry,
+        ExtensionAttributesFactory $extensionFactory,
+        AttributeValueFactory $customAttributeFactory,
```
Please, inject new dependencies according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/)

deefco replied:
Am I wrong in thinking that I need to make the object extend AbstractExtensibleModel? If so, how do I make these changes backwards compatible, seeing how they require completely different things in the constructor? Just use the object manager for everything?

sidolov commented on app/code/Magento/Sitemap/Model/Sitemap.php:
```diff
@@ -278,14 +282,6 @@ protected function _getStream()
-     * @return $this
-     * @deprecated 100.2.0
-     * @see ItemProviderInterface
-     */
```
Can you explain why this doc block was removed?

deefco replied:
Replaced

sidolov commented on app/code/Magento/Sitemap/Model/Sitemap.php:
```diff
@@ -362,12 +354,6 @@ protected function _initSitemapItems()
-     *
-     * @return \Magento\Framework\Model\AbstractModel
-     * @throws LocalizedException
-     */
```
Can you explain why this doc block was removed?

deefco replied:
Replaced

sidolov commented on app/code/Magento/Sitemap/Model/Sitemap.php:
```diff
@@ -750,13 +736,9 @@ protected function _getStoreBaseDomain()
+     * {@inheritdoc}
      */
-    public function getSitemapUrl($sitemapPath, $sitemapFileName)
+    public function getSitemapUrl($sitemapPath, $sitemapFileName): ?string
```
We are not allowed to change public methods according to our [Backward Compatible Development Guide](https://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/)

deefco replied:
Removed getSitemapUrl from interface because it is not a clean dto method but has side effects.

sidolov commented on app/code/Magento/Sitemap/Model/SitemapItem.php:
```diff
@@ -90,4 +90,17 @@ public function getUpdatedAt()
         return $this->updatedAt;
     }
+
+    public function toArray()
```
Please, clarify the purpose of this method, looks like it never used

deefco replied:
Removed

deefco committed:
Remove SitemapExtensionInterface and make SitemapInterface extend ExtensionAttributesInterface

deefco committed:
Call delete instead of save in delete() call

deefco committed:
Replace removed docblocks

deefco committed:
Remove unused toArray function

deefco committed:
Remove getSitemapUrl from interface

deefco committed:
Fix b/c issues in constructor

deefco commented:
@magento-engcom-team give me test instance

deefco commented:
@magento-engcom-team give me test instance
